### ISSUE TITLE
[makeotf] GDEF should not contain GlyphClassDef entries with class 0

### ISSUE THREAD

**Issue by brawer** [adobe-type-tools/afdko#89]
For the following feature file, makeotf v2.0.90 (Nov 19 2015) generates a GDEF table with surprising GlyphClassDef entries:
```
languagesystem DFLT dflt;

@A_MARKS = [acute grave macron];
@B_MARKS = [ogonek];
markClass [caron cedilla] <anchor 350 0> @C_MARKS;

lookup A {
    lookupflag MarkAttachmentType @A_MARKS;
    pos A 1;
} A;

lookup B {
    lookupflag MarkAttachmentType @B_MARKS;
    pos B 1;
} B;

lookup C {
    lookupflag MarkAttachmentType @C_MARKS;
    pos C 1;
} C;

lookup D {
    lookupflag RightToLeft IgnoreBaseGlyphs MarkAttachmentType @C_MARKS;
    pos D 1;
} D;

lookup E {
    lookupflag RightToLeft;
    pos E 1;
} E;

lookup F {
    lookupflag IgnoreBaseGlyphs;
    pos F 1;
} F;

lookup G {
    lookupflag IgnoreLigatures;
    pos G 1;
} G;

lookup H {
    lookupflag IgnoreLigatures;
    pos H 1;
} H;

lookup I {
    lookupflag 6;
    pos I 1;
} I;

feature test {
    lookup A;
    lookup B;
    lookup C;
    lookup D;
    lookup E;
    lookup F;
    lookup G;
    lookup H;
    lookup I;
} test;
```

Observed output:
```
  <GDEF>
    <Version value="1.0"/>
    <GlyphClassDef>
      <ClassDef glyph="caron" class="3"/>
      <ClassDef glyph="cedilla" class="3"/>
      <ClassDef glyph="hungarumlaut" class="0"/>
      <ClassDef glyph="ogonek" class="0"/>
    </GlyphClassDef>
    <MarkAttachClassDef>
      <ClassDef glyph="acute" class="1"/>
      <ClassDef glyph="caron" class="3"/>
      <ClassDef glyph="cedilla" class="3"/>
      <ClassDef glyph="grave" class="1"/>
      <ClassDef glyph="macron" class="1"/>
      <ClassDef glyph="ogonek" class="2"/>
    </MarkAttachClassDef>
  </GDEF>
```

Expected:

* The produced GDEF table should not have a GlyphClassDef entry for `hungarumlaut`. This glyph was not even mentioned in the input feature file.

* The produced GDEF table should not contain any GlyphClassDef entries with class 0. According to the [GDEF specification](https://www.microsoft.com/typography/otspec/gdef.htm), “[t]he font developer does not have to classify every glyph in the font, but any glyph not assigned a class value falls into Class zero (0).” Thus, since the default is zero, there is no need to emit entries with class 0.

Commands to reproduce on MacOS X 10.11.1:
```
$ md5 Skia.ttf 
MD5 (Skia.ttf) = 13711112e6da2dda19bdeee76c52c91d
$ makeotf -f Skia.ttf -ff lookupflag.fea -o out.ttf ; ttx  -t GDEF out.ttf ; cat out.ttx
```

**Comment by readroberts** [comment]
Clearly another bug to fix. I hope to work on these next week.

**Comment by behdad** [comment]
> Clearly another bug to fix. I hope to work on these next week.

Or, switch to the fontTools.feaLib which should be ready by then :).

**Comment by readroberts** [comment]
Having looked at this, I see that the glyph class 0 entries are the result of a file size optimization. makeotf correctly builds a list of glyphs and their assigned glyph class index. makeotf then has to write a ClassDef to represent the set  of glyphs. In your example, it could represent only the specified glyphs with ClassDef2, which supports ranges. It chooses ClassDef1, which is smaller, but requires specifying the class for all glyphs in the glyphID range from min GID to max GID. This means that a bunch of glyphs get mapped to class 0. This is ok: the spec says that you do not have to map glyphs to class 0; it does not say that you cannot do so. I think this would be a bug if the glyphs were incorrectly mapped to a class ID, but that is not the case here.

**Comment by behdad** [comment]
Right.  That's a legitimate optimization.  Sascha, would be nice to compare
the feaLib output size against makeotf for this case.  FWIW, the respective
code in fonttools is in otTables.py:ClassDef.preWrite(), if you want to
improve the heuristic.

On Tue, Apr 26, 2016 at 4:39 PM, Read Roberts <notifications@github.com>
wrote:

> Closed #89 <https://github.com/adobe-type-tools/afdko/issues/89>.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/adobe-type-tools/afdko/issues/89#event-642234111>
>



-- 
behdad
http://behdad.org/
